File: app/src/main/java/com/example/pc/basemvp/data/source/RepositoryModule.java
package com.example.pc.basemvp.data.source;

import com.example.pc.basemvp.data.source.local.sharePrf.SharedPrefsApi;
import com.example.pc.basemvp.data.source.remote.api.middleware.InterceptorImpl;
import com.example.pc.basemvp.repositories.TokenRepository;
import com.example.pc.basemvp.repositories.UserRepository;
import com.example.pc.basemvp.repositories.impl.TokenRepositoryImpl;
import com.example.pc.basemvp.repositories.impl.UserRepositoryImpl;
import dagger.Module;
import dagger.Provides;
import javax.inject.Singleton;

@Module
public class RepositoryModule {

    @Provides
    @Singleton
    public TokenRepository provideTokenRepository(SharedPrefsApi sharedPrefsApi) {
        return new TokenRepositoryImpl(sharedPrefsApi);
    }

    @Provides
    @Singleton
    public UserRepository provideUserRepository(TokenRepository tokenRepository) {
        InterceptorImpl.getInstance().setTokenRepository(tokenRepository);
        return new UserRepositoryImpl();
    }
}

File: flow_hilt/app/src/main/java/com/example/flowmvvm/di/RepositoryModule.kt
package com.example.flowmvvm.di

import android.app.Application
import com.example.flowmvvm.data.source.local.dao.AppDatabase
import com.example.flowmvvm.data.source.local.sharedprf.SharedPrefsApi
import com.example.flowmvvm.data.source.local.sharedprf.SharedPrefsImpl
import com.example.flowmvvm.data.source.remote.service.ApiService
import com.example.flowmvvm.data.source.repositories.*
import com.example.flowmvvm.utils.dispatchers.BaseDispatcherProvider
import com.google.gson.Gson
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object RepositoryModule {

    @Singleton
    @Provides
    fun provideTokenRepository(app: Application): TokenRepository {
        return TokenRepository(SharedPrefsImpl(app))
    }

    @Singleton
    @Provides
    fun provideAppDBRepository(
        dispatcherProvider: BaseDispatcherProvider,
        appDatabase: AppDatabase,
        gson: Gson
    ): AppDBRepository {
        return AppDBRepositoryImpl(dispatcherProvider, appDatabase, gson)
    }

    @Singleton
    @Provides
    fun provideUserRepository(
        dispatcherProvider: BaseDispatcherProvider,
        apiService: ApiService,
        sharedPrefsApi: SharedPrefsApi,
        gson: Gson
    ): UserRepository {
        return UserRepositoryImpl(dispatcherProvider, apiService, sharedPrefsApi, gson)
    }
}

